From 5f2a7d6811d47f1a241333563cd853b0f6256dfd Mon Sep 17 00:00:00 2001
From: Andrey Yepin <ayepin@google.com>
Date: Thu, 22 May 2025 23:09:24 -0700
Subject: [PATCH 1/2] Sanitize cross-profile intents.

Remove package or component information from payload intents (and their
selectors) for cross-profile sharing.

Bug: 407764858
Test: manual testing with test app
Flag: EXEMPT bugfix
Merged-In: I1ebfd96b2aabba6665267722603c72cbe4aefe0f
Change-Id: Ie00f0a4e2fdb0cb7a3b2a148369d435d7da9b1bc
---
 .../intentresolver/ChooserActivity.java       | 10 ++++-
 .../intentresolver/util/IntentUtils.kt        | 41 +++++++++++++++++++
 2 files changed, 49 insertions(+), 2 deletions(-)
 create mode 100644 java/src/com/android/intentresolver/util/IntentUtils.kt

diff --git a/java/src/com/android/intentresolver/ChooserActivity.java b/java/src/com/android/intentresolver/ChooserActivity.java
index 3ce21e29..15bb1e06 100644
--- a/java/src/com/android/intentresolver/ChooserActivity.java
+++ b/java/src/com/android/intentresolver/ChooserActivity.java
@@ -24,6 +24,7 @@ import static android.app.admin.DevicePolicyResources.Strings.Core.RESOLVER_CROS
 import static android.stats.devicepolicy.nano.DevicePolicyEnums.RESOLVER_EMPTY_STATE_NO_SHARING_TO_PERSONAL;
 import static android.stats.devicepolicy.nano.DevicePolicyEnums.RESOLVER_EMPTY_STATE_NO_SHARING_TO_WORK;
 
+import static com.android.intentresolver.util.IntentUtils.sanitizePayloadIntents;
 import static com.android.internal.util.LatencyTracker.ACTION_LOAD_SHARE_SHEET;
 
 import android.annotation.IntDef;
@@ -516,9 +517,12 @@ public class ChooserActivity extends ResolverActivity implements
             boolean filterLastUsed,
             TargetDataLoader targetDataLoader) {
         int selectedProfile = findSelectedProfile();
+        List<Intent> crossProfileIntents = sanitizePayloadIntents(mIntents);
         ChooserGridAdapter personalAdapter = createChooserGridAdapter(
                 /* context */ this,
-                /* payloadIntents */ mIntents,
+                /* payloadIntents */ selectedProfile == PROFILE_PERSONAL
+                        ? mIntents
+                        : crossProfileIntents,
                 selectedProfile == PROFILE_PERSONAL ? initialIntents : null,
                 rList,
                 filterLastUsed,
@@ -526,7 +530,9 @@ public class ChooserActivity extends ResolverActivity implements
                 targetDataLoader);
         ChooserGridAdapter workAdapter = createChooserGridAdapter(
                 /* context */ this,
-                /* payloadIntents */ mIntents,
+                /* payloadIntents */ selectedProfile == PROFILE_WORK
+                        ? mIntents
+                        : crossProfileIntents,
                 selectedProfile == PROFILE_WORK ? initialIntents : null,
                 rList,
                 filterLastUsed,
diff --git a/java/src/com/android/intentresolver/util/IntentUtils.kt b/java/src/com/android/intentresolver/util/IntentUtils.kt
new file mode 100644
index 00000000..5c909677
--- /dev/null
+++ b/java/src/com/android/intentresolver/util/IntentUtils.kt
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@file:JvmName("IntentUtils")
+
+package com.android.intentresolver.util
+
+import android.content.Intent
+
+fun sanitizePayloadIntents(intents: List<Intent?>): List<Intent?> =
+    intents.map { intent ->
+        if (intent == null) {
+            null
+        } else {
+            Intent(intent).also { sanitized ->
+                sanitized.setPackage(null)
+                sanitized.setComponent(null)
+                sanitized.selector?.let {
+                    sanitized.setSelector(
+                        Intent(it).apply {
+                            setPackage(null)
+                            setComponent(null)
+                        }
+                    )
+                }
+            }
+        }
+    }
-- 
2.34.1

